### DISCUSSION TITLE
C++ pow with const error

### DISCUSSION THREAD

**Discussion by oldercoder** [libvips/libvips#2741] [Q&A]
Hi there,

I'm writing C++ code by using libvips. I got a very strange error on VImage pow calculation, blow is the simple code and error:

```C++
    const char *filename = "C:\\img\\test1.png";
    auto in = vips::VImage::new_from_file(filename);
    if (in.is_null()) {
        printf("Can not read image, quiting...");
        return -1;
    }
    auto out = in.pow(2.0); // this line will cause the compiler error on below:
```

Error:

```
C:\Program Files\JetBrains\CLion 2020.2.3\bin\mingw\bin/ld.exe: CMakeFiles/app1.dir/app1.cpp.obj: in function `vips::VImage::pow(double, vips::VOption*) const':
D:/C++libs/vips-dev-8.12/include/vips/VImage8.h:1484: undefined reference to `vips::VImage::math2_const(VipsOperationMath2, std::vector<double, std::allocator<double> >, vips::VOption*) const'
```

I'm using mingGW64 compiler on windows, any idea would be great, thanks.

**Comment by jcupitt** [comment]
Hi @oldercoder,

That's odd. I made a test program:

```C++
/* compile with:
 *
 *      g++ -g -Wall pow.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips8>

int
main (int argc, char** argv)
{
    VIPS_INIT(argv[0]);

    auto in = vips::VImage::new_from_file(argv[1]);
    auto out = in.pow(2.0); 
    out.write_to_file(argv[2]);

    return 0;
}
```

And it's fine for me (on Ubuntu with git master libvips). If that works for you, I suppose there must be some kind of install problem.

So .... what version of libvips did you install, and how?

**Reply by jcupitt** [reply]
(you don't need to check the result of `new_from_file()`, it'll throw an error if the file is not found)

**Reply by oldercoder** [reply]
Hi @jcupitt , thanks for quick test.
I did test the program on my Mac, it's works fine on that. The reason I asked is when I move to windows, the same program won't compile due to above error.
My windows environments is : mingw g++ (GCC) 11.2.0, with libvips is latest one v8.12.2. It seems can't find math2_const method, but others method are all ok, it's strange. During further investigation, I found "get_point" method is not work complaining:
```
"warning: HEAP[app1.exe]: 
warning: Invalid address specified to RtlFreeHeap( 0000018669B30000, 000001866B5F15E0 )

Thread 1 received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffcb5c4a313 in ntdll!RtlRegisterSecureMemoryCacheCallback () from C:\Windows\SYSTEM32\ntdll.dll"
```

it's very strange since very thing is fine and work on my Mac...

**Reply by jcupitt** [reply]
It's probably a g++ issue --- you can't mix C++ compilers in a project, since C++ has no ABI. You *must* use exactly the same C++ compiler (and the same version of that compiler!) for every library in your program.

We build the libvips binary with this thing:

https://github.com/libvips/build-win64-mxe

So you'll need to either match that g++ version, or rebuild the C++ API using your own C++ compiler (quite easy), or use the C interface.

To rebuild the libvips C++ API, just copy this small directory tree into your project:

https://github.com/libvips/libvips/tree/master/cplusplus

And build it along with your code.

**Reply by oldercoder** [reply]
Thanks for advice, jcupitt, I tried C interface, it's worked on pow now! below is the code I used:
```
vips::VImage pow_vip(const vips::VImage &in, double n) {
    VipsImage *i, *o;
    vips::VImage out;

    i = in.get_image();
    vips_pow_const1(i, &o, n, NULL);
    out = o;

    return out;
}
```
I'm working on get_point now, thanks again for help! You really give me the direction to solve problem.

**Reply by jcupitt** [reply]
I'm not sure, that might refleak. I would just build the official C++ API, you'll save a lot of time.

I'm glad it's working anyway.
